src/proxy/outbound.rs

```diff
@@ -380,6 +381,64 @@ impl OutboundConnection {
             None => return Err(Error::UnknownSource(downstream)),
         };
 
+        // If this is to-service traffic check for a service waypoint
+        if let Some(Address::Service(s)) = self
+            .pi
+            .state
+            .fetch_destination(&Destination::Address(NetworkAddress {
+                network: self.pi.cfg.network.clone(),
+                address: target.ip(),
+            }))
+            .await
+        {
+            // if we have a waypoint for this svc, use it; otherwise route traffic normally
+            if let Some(wp) = s.waypoint.clone() {
+                let waypoint_vip = match wp.destination {
+                    Destination::Address(a) => a.address,
+                    Destination::Hostname(_) => {
+                        return Err(proxy::Error::UnknownWaypoint(
+                            "hostname lookup not supported yet".to_string(),
+                        ));
+                    }
+                };
+                let waypoint_vip = SocketAddr::new(waypoint_vip, wp.hbone_mtls_port);
+                let waypoint_us = self
+                    .pi
+                    .state
+                    .fetch_upstream(&self.pi.cfg.network, waypoint_vip)
+                    .await
+                    .ok_or(proxy::Error::UnknownWaypoint(
+                        "unable to determine waypoint upstream".to_string(),
+                    ))?;
+
+                let waypoint_workload = waypoint_us.workload;
+                let waypoint_ip = self
+                    .pi
+                    .state
+                    .load_balance(
+                        &waypoint_workload,
+                        &source_workload,
+                        self.pi.metrics.clone(),
+                    )
+                    .await?; // if we can't load balance just return the error
+
+                let waypoint_socket_address = SocketAddr::new(waypoint_ip, waypoint_us.port);
+
+                return Ok(Request {
+                    protocol: Protocol::HBONE,
+                    direction: Direction::Inbound,
```

Comment:
isn't it outbound?

Reply:
probably should be. I did a copy/paste to start building the request and it seems to be a hangover from a hangover.

https://github.com/istio/ztunnel/pull/486/files#r1187432943